### PR TITLE
Respect preexisting file ACLs in overwrite YAML utility routine

### PR DESCRIPTION
**Pull request by flongford**
## Related Issues/PRs

#8878 

## What changes are proposed in this pull request?

The ACLs of the preexisting yaml are copied over to the temporary file via `shutil.copystat` before it is moved to replace the preexisting file in `overwrite_yaml`.

By doing so we retain the atomicity of the update operation (although `shutil.move` is not _purely_ atomic anyway), so that the file permissions of the preexisting yaml remain unchanged throughout.

A regression test is also added that will fail on `master`.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Bug fix for `PermissionError` exceptions reported by multiple users in a NFS-backed tracking store after experiments are renamed or deleted. Solution ensures that ACLs of preexisting experiment `meta.yaml` files are respected when performing updates.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/utils/test_file_utils.py [hunk 1]
```diff
@@ -363,3 +363,19 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
...
+    file_utils.write_yaml(tmp_dir, yaml_path.name, {"foo": "bar"})
+    expected_mode = 0o100660
+    # ensure temporary file does not have elevated permissions
+    os.chmod(str(yaml_path), expected_mode)
+
+    # When
+    file_utils.overwrite_yaml(tmp_dir, yaml_path.name, {"bar": "foo"})
+
+    # Then
+    assert os.stat(str(yaml_path)).st_mode == expected_mode
```
Comment (harupy) [inline tests/utils/test_file_utils.py:381]:
```suggestion
def test_overwrite_yaml_preserves_preexisting_permissions(tmp_path):
    tmp_dir = str(tmp_path)
    yaml_path = tmp_path / random_file("yaml")
    file_utils.write_yaml(tmp_dir, yaml_path.name, {"foo": "bar"})
    expected_mode = 0o100660
    # ensure temporary file does not have elevated permissions
    os.chmod(str(yaml_path), expected_mode)
    file_utils.overwrite_yaml(tmp_dir, yaml_path.name, {"bar": "foo"})
    assert os.stat(str(yaml_path)).st_mode == expected_mode
```

Do we need these comments?

Comment (harupy) [reply tests/utils/test_file_utils.py:381]: Let's remove them :)

Comment (flongford) [reply tests/utils/test_file_utils.py]: Done in [0c5720f](https://github.com/mlflow/mlflow/pull/8899/commits/0c5720f462dfb3d2ba5107941677ef2806801fef)

File: tests/utils/test_file_utils.py [hunk 2]
```diff
@@ -363,3 +363,14 @@ def test_shutil_copytree_without_file_permissions(tmp_path):
...
+
+def test_overwrite_yaml_preserves_preexisting_permissions(tmp_path):
+    tmp_dir = str(tmp_path)
+    yaml_path = tmp_path / random_file("yaml")
+    file_utils.write_yaml(tmp_dir, yaml_path.name, {"foo": "bar"})
+    expected_mode = 0o100660
+    # ensure temporary file does not have elevated permissions
+    os.chmod(str(yaml_path), expected_mode)
+    file_utils.overwrite_yaml(tmp_dir, yaml_path.name, {"bar": "foo"})
+    assert os.stat(str(yaml_path)).st_mode == expected_mode
```
Comment (harupy) [inline tests/utils/test_file_utils.py:376]:
```suggestion
    assert os.stat(yaml_path).st_mode == expected_mode
```

Comment (harupy) [inline tests/utils/test_file_utils.py:376]: /ai can this code be simplified if `yaml_path` is a `pathlib.Path` object?

Comment (flongford) [reply tests/utils/test_file_utils.py]: Changed in [a2d0f66](https://github.com/mlflow/mlflow/pull/8899/commits/a2d0f66298885015b964d0c623a7df3371d3c3d4)

**Comment by flongford** [comment]
@harupy - the unit tests ran on [a2d0f66](https://github.com/mlflow/mlflow/pull/8899/commits/a2d0f66298885015b964d0c623a7df3371d3c3d4) were failing because it turns out `os.chmod` is only partially supported for Windows: https://docs.python.org/3/library/os.html#os.chmod. 

Therefore, the following command to set the testing file permissions to a known value had no effect on a Windows build:

https://github.com/mlflow/mlflow/blob/a2d0f66298885015b964d0c623a7df3371d3c3d4/tests/utils/test_file_utils.py#L372-L373

I've updated the test in [aa5db46](https://github.com/mlflow/mlflow/pull/8899/commits/aa5db46d23966d36f4517a7451643ac6ed1ec202) to use a value that should work cross-platform on a GitHub Actions runner.

However, after some investigation, I'm starting to wonder whether all files (even "secure" temporary) ones created on Windows _do_ seem to respect the global environment ACLs on Windows. If so, this "bug" might only affect Unix backends. In which case the latest incarnation of the unit test _might_ fail on a local development environment with different ACLs.

I will do a little more investigating to see what the best approach should be here.

_(disclaimer: I'm don't develop in Windows day-to-day, so setting up an appropriate environment to test this out will take a bit of time)_

**Comment by flongford** [comment]
> I will do a little more investigating to see what the best approach should be here.

@harupy - I've improved the robustness of the unit test in [e15f733](https://github.com/mlflow/mlflow/pull/8899/commits/e15f733ec47d7d796ef840090cbc1a6eb28d4495) and [244cac7](https://github.com/mlflow/mlflow/pull/8899/commits/244cac7247437906bfe6dd610054a3a1037be359), considering:
- It doesn't appear that we can reliably set the permissions on the pre-existing yaml file across all supported OS
- Therefore, we can only check that whatever permissions `file_utils.write_yaml` sets on the pre-existing file are retained after performing `file_utils.overwrite_yaml`
- If the pre-existing file is created with elevated permissions (specifically 600), then the test will fail to detect regressions (in fact this test would pass on `main` in that scenario). Therefore, I've added a warning to the test in the event that this occurs.

**Comment by flongford** [comment]
Closing, since the original issue reported appears to have been resolved by https://github.com/mlflow/mlflow/pull/9799

**Pull request closed without merging**